Add Perforce as supported VCS

barneygale opened the pull request:
This PR is a first pass on supporting the Perforce VCS, as described in #2754.

Still to-do:

- [x] Better client name generation
- [x] Testing on Windows and Mac
- [x] Automated tests

Notes:

- `controls_location()` always returns `False`, and hence installations from the local filesystem are impossible. As such every installation is a fresh sync from the remote depot. I *think* this makes sense for perforce - keen to hear usecases for anything else.
- `p4 client` won't accept a client specification using a command-line argument. It *can* accept one on stdin, but `pip`'s command-runner doesn't currently support passing data to stdin. For the timebeing I've worked around this by writing the client spec to a temporary file, and then setting `P4EDITOR` to copy that file. Happy to consider other approaches.

I hope this is along the right lines?

pradyunsg commented on news/2754.feature:
```diff
@@ -0,0 +1,2 @@
+Add Perforce as a supported VCS. Packages can be installed from the perforce
```
nit: add a newline.

```
pip can now install packages from the perforce depot as ``pip install p4+p4://depot/path/to/package``.
```

barneygale commented:
Thanks for taking a look @pradyunsg. I will fix the niggle in the next revision of this patch.

I think tests will require installations of `p4` (client) and `p4d` (server) on the test boxes. These can be downloaded from the perforce website free-of-charge. I'm not sure if the server (besides the executable) needs to be persistent - it could probably be spun up/down in a pytest fixture. An alternative would be to test against a public perforce server, but the tests probably shouldn't rely on having a working internet connection right?

barneygale commented:
A couple more questions for someone with p4 knowledge

1. Is it reasonable for `pip install p4+p4://` to accept only a depot path, and not a server host/port or username/password? Currently the [`P4PORT`](https://www.perforce.com/perforce/doc.current/manuals/cmdref/#CmdRef/P4PORT.html), [`P4USER`](https://www.perforce.com/perforce/doc.current/manuals/cmdref/#CmdRef/P4USER.html) and [`P4PASSWD`](https://www.perforce.com/perforce/doc.current/manuals/cmdref/#CmdRef/P4PASSWD.html) environment variable is passed through unchanged, which squares with my normal usage of perforce (I set these in my `~/.bashrc`). 
2. Previously both the company and the product were called "Perforce". In the last few years they've rebranded the product as "Helix" or "Helix Core", but the command-line utilities and whatnot are still called "p4", and as far as I can tell everyone still calls it "perforce". I'm not sure whether in `pip` we want to mirror official usage (i.e. use "Helix" in docs but "p4" in the implementation) or use Perforce/p4 consistently.

pradyunsg commented:
The tests already depend on the network. I'd prefer offline tests though but I don't know how much work it would be to setup stuff in a session wide fixture. 

I obviously defer to your or the reviewer's judgement here.

> questions for someone with p4 knowledge

Not me. :P

pfmoore commented:
As noted on #2754, we'd need someone to maintain this code. @barneygale are you likely to be able to handle support going forward? Otherwise the code is likely to bitrot.

As for tests, we'd need tests in Travis and Appveyor, so you'll need to do something in the config to automate the download and install of Perforce as part of the test run. I'd prefer tests not to rely on a public Perforce repo, as if the chosen site goes down, we'd have all our test runs failing until a fix could be found.

barneygale commented:
I'm happy to maintain this going forward.

Tests are in-progress; my current approach is to mirror the tests for other VCSs, and to set up a local server. I'll make sure travis/appveyor configs are updated to get `p4`/`p4d` installed.

barneygale commented:
I've pushed various improvements, most notably that the URL format can now contain information that's translated into P4PORT, i.e. protocol, hostname, port. If none of these are specified, P4PORT is passed through unchanged. I think this covers two usecases:

- Installing within corporate environments where P4PORT always points to the same server on the LAN
- Installing from internet-facing servers, where you need to be able to put all relevant info in a 
  'requirements.txt' or 'setup.py'.

I've also added a functional test that runs `pip install ...`. There's a new session-scoped fixture to spin up a perforce server in a temporary directory. 

I'm not sure what's needed to get the `p4` and `p4d` executables on to travis and appveyor. Is there a file in the repository to edit?

pradyunsg commented:
Travis build setup happens in .travis/install.sh, Appveyor build setup happens in appveyor.yml (if that's what you're asking)

barneygale commented:
Thanks. I ask because I can't see any stuff in those files to install other VCS (bazaar, mercurial, etc). It looks like travis takes care of Linux and appveyor takes care of Windows? My plan is to edit those scripts to download `p4` and `p4d` from the perforce website and add them to $PATH.

pradyunsg commented:
> It looks like travis takes care of Linux and appveyor takes care of Windows?

Yep. 

- https://docs.travis-ci.com/user/reference/trusty/#Version-control
- https://www.appveyor.com/docs/build-environment/#version-control-systems

barneygale commented:
Travis configuration changes seem to be working, appveyor is on the way.

I still need to test what happens if `p4` attempts to interactively prompt the user to log in.

Remaining questions:

1. Is a single `pip install p4+...` integration test sufficient? I can write further tests, and I would appreciate some guidance on what to cover.
2. Should it be possible to specify the perforce username/password via the URL? We could support something along the lines of `p4+p4://username:password@host:port//depot/path`. I hesitate for a couple of reasons:
  a. For security reasons web browsers don't support these form of URL any more
  b. The `host:port` bit is currently optional (if omitted, `P4PORT` is passed through unchanged). This would result in slightly freaky-looking URLs like `p4+p4://username:password@//depot/path` if user/password should be set, but not host/port.
3. Are we happy referring to this as "Perforce" rather than "Helix"?

barneygale commented:
... and sorry for DOSing the CI. I'm not sure if you normally squash when you merge PRs - if not I can split the final PR diff into a small handful of commits.

pradyunsg commented:
1. No? I mean, are there multiple situations like git can have? (like
default branch is not master) If so, just everything that you think is
reasonable to support should get its own test. I think the bzr/svn/git
tests were a good reference.

Aside, if possible add more unit tests. I personally want to push  the unit
test ratio higher in the coming months and I'd appreciate it if you could
help with that. :)

2. I have some questions about this... What are your security concerns? How
common are these? Maybe it makes sense to support them if they're common
enough? Does pip allow this with git?

3. Let's go with whichever the official documentation uses.

As for the CI DOS, I don't see much other activity around here right now
so, it's cool. No one was affected by the "outage". :P

On Mon, 4 Dec 2017, 01:30 Barney Gale, <notifications@github.com> wrote:

> Travis configuration changes seem to be working, appveyor is on the way.
>
> I still need to test what happens if p4 attempts to interactively prompt
> the user to log in.
>
> Remaining questions:
>
>    1. Is a single pip install p4+... integration test sufficient? I can
>    write further tests, and I would appreciate some guidance on what to cover.
>    2. Should it be possible to specify the perforce username/password via
>    the URL? We could support something along the lines of
>    p4+p4://username:password@host:port//depot/path. I hesitate for a
>    couple of reasons:
>    a. For security reasons web browsers don't support these form of URL
>    any more
>    b. The host:port bit is currently optional (if omitted, P4PORT is
>    passed through unchanged). This would result in slightly freaky-looking
>    URLs like p4+p4://username:password@//depot/path if user/password
>    should be set, but not host/port.
>    3. Are we happy referring to this as "Perforce" rather than "Helix"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/4857#issuecomment-348809666>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADH7SQyxZWWuNNF04EKIaJE1L4aZnHlkks5s8v3LgaJpZM4QaoK1>
> .
>

benoit-pierre commented on appveyor.yml:
```diff
@@ -23,6 +23,12 @@ install:
   - "set GIT_SSL_CAINFO"
+  # Install perforce.
+  - "mkdir %TEMP%\\perforce\\bin"
+  - "pushd %TEMP\\perforce\\bin"
```
I think this line is missing a `%`.

benoit-pierre commented on appveyor.yml:
```diff
@@ -23,6 +23,12 @@ install:
+  - "mkdir %TEMP%\\perforce\\bin"
+  - "pushd %TEMP\\perforce\\bin"
+  - "appveyor DownloadFile http://ftp.perforce.com/perforce/r17.2/bin.ntx86/p4.exe"
+  - "appveyor DownloadFile http://ftp.perforce.com/perforce/r17.2/bin.ntx86/p4d.exe"
```
I don't know how big the downloads are, but maybe installing Perforce should be conditional on running the integration tests.

benoit-pierre commented:
@pradyunsg: is support for [rolling builds](https://www.appveyor.com/docs/build-configuration/#rolling-builds) enabled on AppVeyor?

barneygale commented:
Hi all,

I've added support for other VCS operations like `update()`, `switch()` and `export()`. As perforce doesn't have an equivalent to a `.git` or `.svn` directory, I've created a `.p4pip` convention to support this.

I've also added a couple of unit tests for the new `split_url` and `join_url` functions. The rest of the code is fairly stateful, so I could expand the functional tests to hit the `update()`, `switch()` and `export()` code paths. I couldn't see equivalent functional tests for other VCSs, if that matters.

> > Should it be possible to specify the perforce username/password via the URL?
> 
> I have some questions about this... What are your security concerns? How common are these? Maybe it makes sense to support them if they're common enough? Does pip allow this with git?

After a bit of googling it seems that pip does support this with other VCSs, so I've added support here. As with pip's subversion support, the URL is 'canonicalized' before being displayed back to the user, so the username, password, protocol, revision and fragment are omitted.

> Let's go with whichever the official documentation uses.

I've switched to using 'Helix Core'

I'm pretty happy with the code/layout at this stage. Next steps from me are a few more functional tests, getting CI passing, and handle any review comments from you folks.

Cheers

pradyunsg commented:
> @pradyunsg: is support for rolling builds enabled on AppVeyor?

It should be but I'm actually not sure. @pfmoore or @dstufft would have to check. I'm not an admin on Appveyor.

pradyunsg commented:
@barneygale could you add WIP in the title of this PR and remove it once is not longer WIP?

Assuming, it's still a WIP. :)

barneygale commented:
Have done so! I wanted to share for some early feedback on the approach. Last remaining task is now to expand the functional tests, after which I believe I can remove the WIP tag.

barneygale commented:
I believe this patch is now ready to review. The history is a bit of a mess as I was learning both pip and p4 along the way :-)

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

barneygale commented:
Hey folks, anything I can do to make this more reviewable? I can put in a new PR if the history is a bit too much of a mess (though I suppose you'll squash-merge this anyway)

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

barneygale commented:
Rebased again. Let me know if writing up more information on this patch would make it more reviewable.

pradyunsg commented:
Hey! Sorry, I'd missed that in the swamp of notifications.

One option usually is to break up a monolithic PR into smaller individually coherent commits or even separate PRs.

I'll try to come around to this but I'll wait for at least one more reviewer before merging.

pradyunsg commented:
Whoops! I shouldn't fiddle around before waking up properly.

barneygale commented:
Rewrote the history to split by area:

- 8895fba is the main implementation
- 9f24a7a updates documentation and news
- 2f50d11 updates the appveyor + travis configuration to install the perforce client + server
- 653a4a3 add functional + unit tests

barneygale commented:
If it helps, I've found that [Perforce in FreeBSD Development](https://www.freebsd.org/doc/en/articles/p4-primer/) as a good primer for Perforce. Perforce is quite a lot like subversion, except that:

- Perforce (the company) provide commercial support for Helix Core (the product formerly known as Perforce). This is attractive for larger organisations.
- Perforce is thought to be better at handling large binary files than other VCSs. This is one of the reasons it gets used in the games + film industries
- On a technical level:
  - Perforce can operate over several network transports. This is determined by [`P4PORT`](https://www.perforce.com/perforce/doc.current/manuals/cmdref/#CmdRef/P4PORT.html)
  - Authentication is entirely username/password based
  - The perforce server keeps a record of user "clients" (or "workspaces"; as far as I can tell the terms are synonymous). Clients are identified by name, and the name must be globally unique across the server. Clients record a mapping from a local filesystem path to a remote "depot" path. A user typically creates several clients on one machine to work with different parts of the depot. Clients can be created/edited/deleted with the `p4 client` command.
  - By default, a checkout doesn't record any information on the filesystem to indicate it's a checkout (unlike `.svn`, `.git`, etc). In common use, users set the [`P4CONFIG`](https://www.perforce.com/perforce/doc.current/manuals/cmdref/index.html#CmdRef/P4CONFIG.html) environment variable to `.p4config`, and then drop `.p4config` files in their checkouts. This file contains something like `P4CLIENT=name_of_my_client`.

This patch doesn't use `P4CONFIG`, but rather implements an similar mechanism. When a user checks out a perforce path (e.g. `pip install p4+p4:///depot/path`) pip parses the URL underlays environment variables to determine `P4PORT`, `P4USER`, `P4PASSWD`, the depot path and the revision. This is saved to a `.p4pip/environ.txt` file, which can be subsequently consulted for implementing `update()` and `switch()`. Client names are randomly-generated. I've tried various approaches for attempting to hash `P4` variables in to something that would by globally unique, but I don't believe there's such a scheme. We can't clean clients up because `--editable` checkouts need to continue to work.

pfmoore commented:
I'd suggest that another part of the problem is, unfortunately, a simple lack of interest in Perforce support. Personally, I don't object to supporting Perforce, but I don't have any interest in, or need, for such support, so I'm not likely to find time to review or merge this PR. The only other pip committer active on this issue is @pradyunsg, so if he feels similarly, you're a bit stuck.

I don't have a good answer here - clearly pip isn't built to solely satisfy the needs of the core committers, but equally, we don't simply accept every feature request on a "so what's the harm?" basis. Probably your best way of getting some traction would be to demonstrate a reasonable level of user need for this feature - maybe by starting a discussion on distutils-sig asking for people with an interest to speak up, or perhaps by pointing to organisations or larger projects that you know would use the feature if it were available. 

My personal view is that VCS support is something that would be far better handled via 3rd party plugins, so that something like `pip-p4` could be developed and maintained independently from pip, by users who have a need for and interest in it. Unfortunately, pip doesn't *have* a plugin mechanism, nor does it have a supported API that plugins could use, so there's a chunk of infrastructure that's needed before that's an option.

barneygale commented:
Understood. There are a few people interested in support in #2754. It's worth noting that perforce does have a fair bit of corporate use - google used it for years before rolling their own version. I think you can find some other large customers with a bit of googling... but that doesn't mean they're interested in using it with `pip`; I'll make a post to the mailing list to gauge interest.

I was attracted to implementing this in pip itself because of pip's existing support for VCSs - the implementation slots in quite nicely, I think. At some point it might make sense to refactor the VCS support into a separate library, which pip could then vendor.

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -678,9 +679,14 @@ def call_subprocess(cmd, show_stdout=True, cwd=None,
+            stdout=stdout,
+            stderr=subprocess.STDOUT,
+            cwd=cwd,
+            env=env)
```
Why is the above being reformatted? Is this pip's recommended style? It doesn't look like there is any change, and the previous formatting seems nicer, at least to me.

barneygale replied:
Apologies - this was a bad rebase on my part. My original patch added `stdin=subprocess.PIPE`, and I reformatted as part of that. But since the patch went into review, 1c36fbc6c13cf1a8998ad0ebfeb9022e3508c18b added `stdin=subprocess.PIPE` too, and so my change now amounts to only formatting tweaks. I'll fix this shortly.

cjerdonek commented on src/pip/_internal/vcs/helix_core.py:
```diff
@@ -0,0 +1,286 @@
+        environ.get("P4USER") or
+        environ.get("USER") or
+        environ.get("USERNAME") or
+        '')
```
I would define `get_p4user()` and `get_p4password()` functions.

barneygale replied:
Could you explain why? The code here would be no more self-contained if moved to another function, as it doesn't use any local variables. I could add some sort of `underlay_environ()` function and call that repeatedly, but sadly `P4PORT` can use a number of formats (`port`, `host:port` or `protocol:host:port`).

cjerdonek replied:
It's more testable that way since you can test the logic independent of the surrounding function. It also reduces the cyclotomic complexity of the surrounding function, which already has a lot of branching.

cjerdonek commented on src/pip/_internal/vcs/helix_core.py:
```diff
@@ -0,0 +1,286 @@
+"""
+
+
+def split_url(url, environ=os.environ):
```
Default of `None` is usually better, using the `if environ is None: ...` pattern.

barneygale replied:
Can you explain more why it's better? I think I'm missing something.

cjerdonek replied:
Two reasons: it's more idiomatic, and it's also friendlier to callers when composing with other functions. To get the default behavior, you can pass None instead of having to duplicate the default value in all calling functions.

cjerdonek commented on src/pip/_internal/vcs/helix_core.py:
```diff
@@ -0,0 +1,286 @@
+    if url.hostname:
+        p4port[1] = url.hostname
+    if url.port:
+        p4port[2] = str(url.port)
```
It seems like it would be more explicit / easily understandable to define `p4port` once you know all the values, instead of defining a candidate list and then mutating the list.

barneygale replied:
There's three contributing layers, from lowest to highest precedence:

1. The (Perforce-defined) defaults: [protocol, host, port]
2. The contents of P4PORT, which can provide [protocol, host, port], [host, port] or [port]
3. The URL, which can provide any combination of these.

I'll play around with the implementation some more and try to make this more explicit.

cjerdonek replied:
I just mean make the final line be `p4port = [scheme, hostname, port]` instead of starting out with default values and then changing the values as you go along. It just seems clearer / more explicit that way.

cjerdonek commented on tests/unit/test_vcs_helix_core.py:
```diff
@@ -0,0 +1,23 @@
+url2 = "p4+ssl://myname:p4ssw0rd@myserver:1234/baz/zyz@456"
+
+
+def test_parse_url():
```
`test_split_url`?

cjerdonek commented on tests/unit/test_vcs_helix_core.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def test_join_url():
+    assert join_url(*split_url(url1, {})[:2]) == "p4://perforce:1666/foo/bar"
```
It seems like it would be better if the test were independent of `split_url()`.

cjerdonek commented on tests/lib/__init__.py:
```diff
@@ -596,6 +596,47 @@ def _vcs_add(script, version_pkg_path, vcs='git'):
+            os.remove(version_pkg_path.join(canary))
+
+            # Delete client
+            client.delete()
```
This seems like a lot of code to be adding to an already long function

cjerdonek commented on tests/lib/__init__.py:
```diff
@@ -553,10 +553,10 @@ def main():
 
 
-def _vcs_add(script, version_pkg_path, vcs='git'):
+def _vcs_add(script, version_pkg_path, vcs='git', p4port=None):
```
Is there a way to structure this function so you don't need to add vcs-specific arguments?

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

pradyunsg commented:
Hey! I'll try to get around to reading this this weekend. It's a lax weekend.

pradyunsg commented:
Hey @barneygale! This slipped under the radar for me. Appreciate the reminder.

The CI doesn't seem too happy about this PR. :/

---

We're currently in the beta period for pip 10.0, which means we won't be merging any new features or enhancements until pip 10.0.0 is released. I've added this to the 10.1 milestone, so that I don't forget about this. :)

Sorry to keep you waiting and I hope you'd understand.

pfmoore commented:
Personally I think it's too niche to warrant including in core pip. I appreciate the time and effort you've put into this, and I'm uncomfortable that we don't have a way of supporting this. I *really* think we need to put some effort into having a VCS plugin API (even if it's surrounded with massive caveats because we don't support general use of the pip API) so that work like this doesn't get wasted. But if I have to take a stance, I'd have to say sorry, but no.

@pypa/pip-committers any comments from others here?

Related: #2523 - Bazaar is another VCS that's probably not really worth core support these days. And the lack of progress on that issue is the sort of problem I worry might happen with Perforce support, too...

pradyunsg commented:
As for whether to include this or not, I feel we should. @barneygale has agreed to help maintain this code moving forward and I think this would be nice as an addition to pip's VCS support.

pradyunsg commented:
Honestly, it makes sense to start supporting plugins for VCS support. It's just that no one has had the time to do so.

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

pradyunsg commented:
Spending some more time looking at code, I just don't feel comfortable say let's merge simply because of all of what @pfmoore says. :/

barneygale commented:
No problem, thanks for considering the proposal and reviewing the code all the same! If I find time I'm going to write some sort of wrapper that monkey-patches perforce support into `pip`. This PR can be closed if everyone is satisfied. Cheers.

RonnyPfannschmidt commented:
i propose a pip entry-point that provides the ability to act based on a vcs url and a target directory

then pip could start to move using it for its own internals,

afterwards moving a implementation to a plugin shouldn't be that hard 

should i open up a new issue with the proposal?

pfmoore commented:
I would love to see some sort of VCS entry point proposal.

RonnyPfannschmidt commented:
im going to hit a night train to go to a vacation rather soon, i#ll tak a look at writing a proposal in that while as well as figuring a explicit documented external facade

barneygale commented:
I got to know some of the VCS code while writing this patch, so I might be useful as a collaborator or reviewer for a VCS entry point proposal/patch. 

It's worth considering whether the current API (`update()`, `export()` etc) is suitable - the SVN code is fine but the git code has plenty of special cases. The abstractions (`VersionControl`, `RevOption`) are a bit wonky and it looks like things have been tacked on for backwards compatibility - which isn't in itself a bad thing. *If* the entry points proposal would make breaking changes, but if it did I reckon it's worth taking another look at the API rather than simply wrapping what's there.

RonnyPfannschmidt commented:
@barneygale its certainly interesting/tricky to sort out ^^

cjerdonek commented:
I have a series of small backwards-incompatible changes to improve the API that I haven't gotten around to proposing as PR's -- to make things more malleable and easier to understand / improve. I made them locally but they became out of date after several style changes were made to pip's code base. I agree that the current API shouldn't be used as is.

cjerdonek commented:
FYI, I dug up the series of incremental commits I made a year ago to simplify the current VCS API. I started applying them to the current master and proposed the first one here: https://github.com/pypa/pip/pull/5529

I can stage additional ones if people are okay with them as they go.

pradyunsg commented:
Closing since there's not much to be done in this PR now. We're not merging since this feels very niche and none of the current pip maintainers are comfortable with this code.

Thanks @barneygale for the PR -- it's extensive and I understand a lot of work went into it -- and everyone else who contributed to the discussion here. :)

@barneygale if you want to help out with other parts of pip, feel free ping me or comment on that issue/PR. :)

pradyunsg commented:
@RonnyPfannschmidt If you don't have enough time now for fleshing out a complete proposal, please do file an issue that can serve as a tracking issue/reminder for all of us. :)

barneygale commented:
Quick question: would it be possible to merge support for passing stdin to child processes, found in [`src/pip/_internal/utils/misc.py`](https://github.com/pypa/pip/pull/4857/files#diff-c9e9f4633cef78355bd4a581febff330) and [`src/pip/_internal/vcs/__init__.py`](https://github.com/pypa/pip/pull/4857/files#diff-20a36458d2828dd6f49cf670a3e81bff)? This would allow me to hack perforce support into `pip` via a wrapper, as a stop-gap until a VCS plugin API is available. If so I can create another PR.

piotr-dobrogost commented:
| There's definitely one along those lines in the tracker somewhere. 

@pradyunsg 
I only managed to find discussion on the topic in issue #2523, see my comment at https://github.com/pypa/pip/issues/5557#issuecomment-404781331
